compiler/src/iree/compiler/Dialect/Flow/Transforms/test/collapse_dimensions.mlir

```diff
@@ -55,5 +56,344 @@ util.func public @unpack_collapse(%arg0: tensor<2x320x128x128xf32>, %arg1: tenso
+#map1 = affine_map<(d0, d1, d2, d3) -> (d1)>
+#map2 = affine_map<(d0, d1, d2, d3) -> (d0, d1)>
+util.func public @prevent_collapse(%arg0: tensor<2x320x128x128xf32>, %arg1: tensor<320xf32>, %arg2: tensor<320xf32>, %arg3: tensor<1x5x2x64xf32>) -> tensor<2x320x128x128xf16> {
+  %0 = flow.dispatch.region -> (tensor<2x320x128x128xf16>) {
+    %1 = tensor.empty() : tensor<2x320xf32>
+    %2 = tensor.empty() : tensor<2x320x128x128xf16>
+    %empty = tensor.empty() : tensor<2x320x128x128xf32>
+    %cst = arith.constant 3.14 : f32
+
+    %elementwise = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%arg0 : tensor<2x320x128x128xf32>) outs(%empty : tensor<2x320x128x128xf32>) {
+    ^bb0(%in : f32, %out : f32):
+      %22 = arith.mulf %cst, %in : f32
+      linalg.yield %22 : f32
+    } -> tensor<2x320x128x128xf32>
+
+    %barrier = util.optimization_barrier %elementwise : tensor<2x320x128x128xf32>
+    %unpack = tensor.unpack %arg3 outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [2, 64] into %1 : tensor<1x5x2x64xf32> -> tensor<2x320xf32>
+
+    %3 = linalg.generic {indexing_maps = [#map, #map1, #map2, #map1, #map], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%barrier, %arg1, %unpack, %arg2 : tensor<2x320x128x128xf32>, tensor<320xf32>, tensor<2x320xf32>, tensor<320xf32>) outs(%2 : tensor<2x320x128x128xf16>) {
+    ^bb0(%in: f32, %in_0: f32, %in_1: f32, %in_2: f32, %out: f16):
+      %4 = arith.addf %in_1, %in_2 : f32
+      %5 = arith.addf %in, %in_0 : f32
+      %6 = arith.truncf %4 : f32 to f16
+      %7 = arith.truncf %5 : f32 to f16
+      %8 = arith.addf %7, %6 : f16
+      linalg.yield %8 : f16
+    } -> tensor<2x320x128x128xf16>
+    flow.return %3 : tensor<2x320x128x128xf16>
+  }
+  util.return %0 : tensor<2x320x128x128xf16>
+}
+
+// CHECK-LABEL:  util.func public @prevent_collapse
+//  CHECK-SAME:    %[[ARG0:.*]]: tensor<2x320x128x128xf32>
+//       CHECK:    %[[ELEMENTWISE:.+]] = linalg.generic
+//  CHECK-SAME:      iterator_types = ["parallel", "parallel", "parallel", "parallel"]
+//  CHECK-SAME:      ins(%[[ARG0]] : tensor<2x320x128x128xf32>)
+//  CHECK-SAME:      outs(%{{.*}} : tensor<2x320x128x128xf32>)
+//       CHECK:    %[[GEN:.+]] = linalg.generic
+//  CHECK-SAME:      iterator_types = ["parallel", "parallel", "parallel", "parallel"]
+//  CHECK-SAME:      ins({{.*}} : tensor<2x320x128x128xf32>, tensor<320xf32>, tensor<2x320xf32>, tensor<320xf32>)
+//  CHECK-SAME:      outs(%{{.*}} : tensor<2x320x128x128xf16>)
+
+// -----
+
+#map = affine_map<(d0, d1, d2) -> (d0, d1, d2)>
+#map1 = affine_map<(d0, d1, d2) -> (d0, d1)>
+#map2 = affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d3, d4)>
+#map3 = affine_map<(d0, d1, d2, d3, d4) -> (d2, d3, d4)>
+#map4 = affine_map<(d0, d1, d2, d3, d4) -> (d0, d1, d2)>
+util.func public @quantized_matmul(%arg0: tensor<4096x32x128xi8>, %arg1: tensor<1x1x32x128xf32>) -> tensor<1x1x4096xf32> {
```

Comment:
I got thrown by this a bit. I guess this doesnt collapse cause of the `linalg.generic` having two `reduction` loops. If it had only one, then the `d0, d1` is collapsible right? Maybe worth adding a test for that.

Reply:
Thank you, I added a test for this and found that I wasn't handling mapping between producer <-> consumer loops. I was just bailing in projected cases more than needed.